sentry-core/src/main/java/io/sentry/core/Hub.java

```diff
@@ -52,165 +54,310 @@ public boolean isEnabled() {
+          options.getLogger(),
+          SentryLevel.WARNING,
+          "Instance is disabled and this 'addBreadcrumb' call is a no-op.");
+    } else if (breadcrumb == null) {
       logIfNotNull(
-          options.getLogger(), SentryLevel.FATAL, "Stack peek was NULL when addBreadcrumb");
+          options.getLogger(), SentryLevel.WARNING, "addBreadcrumb called with null parameter.");
+    } else {
+      StackItem item = stack.peek();
+      if (item != null) {
+        SentryOptions.BeforeBreadcrumbCallback callback = options.getBeforeBreadcrumb();
+        if (callback != null) {
+          breadcrumb = callback.execute(breadcrumb);
+        }
+        if (breadcrumb != null) {
+          item.scope.addBreadcrumb(breadcrumb);
+        }
+      } else {
+        logIfNotNull(
+            options.getLogger(), SentryLevel.FATAL, "Stack peek was null when addBreadcrumb");
+      }
     }
   }
 
+  @NotNull
   @Override
   public SentryId getLastEventId() {
     return lastEventId;
   }
 
   @Override
   public void pushScope() {
-    StackItem item = stack.peek();
-    if (item != null) {
-      Scope clone = null;
-      try {
-        clone = item.scope.clone();
-      } catch (CloneNotSupportedException e) {
-        logIfNotNull(
-            options.getLogger(),
-            SentryLevel.ERROR,
-            "An error has occurred when cloning a Scope",
-            e);
-      }
-      if (clone != null) {
-        StackItem newItem = new StackItem(item.client, clone);
-        stack.push(newItem);
-      }
+    if (!isEnabled()) {
+      logIfNotNull(
+          options.getLogger(),
+          SentryLevel.WARNING,
+          "Instance is disabled and this 'addBreadcrumb' call is a no-op.");
     } else {
-      logIfNotNull(options.getLogger(), SentryLevel.FATAL, "Stack peek was NULL when pushScope");
+      StackItem item = stack.peek();
+      if (item != null) {
+        Scope clone = null;
+        try {
+          clone = item.scope.clone();
+        } catch (CloneNotSupportedException e) {
+          logIfNotNull(
+              options.getLogger(),
+              SentryLevel.ERROR,
+              "An error has occurred when cloning a Scope",
+              e);
+        }
+        if (clone != null) {
+          StackItem newItem = new StackItem(item.client, clone);
+          stack.push(newItem);
+        }
+      } else {
+        logIfNotNull(options.getLogger(), SentryLevel.FATAL, "Stack peek was NULL when pushScope");
+      }
     }
   }
 
   @Override
   public void popScope() {
-    // Don't drop the root scope
-    synchronized (stack) {
-      if (stack.size() != 1) {
-        stack.pop();
+    if (!isEnabled()) {
+      logIfNotNull(
+          options.getLogger(),
+          SentryLevel.WARNING,
+          "Instance is disabled and this 'popScope' call is a no-op.");
+    } else {
+      // Don't drop the root scope
+      synchronized (stack) { // TODO: is it necessary? we should never sync a concurrent object
```

Comment:
@bruno-garcia  do we need that? `stack` is already thread-safe

Reply:
There's a race here:

```
        if (stack.size() != 1) {
          stack.pop();
```